midend/flattenInterfaceStructs.cpp

```diff
@@ -176,4 +204,105 @@ const IR::Node* ReplaceStructs::preorder(IR::P4Control* control) {
     return control;
 }
 
+/////////////////////////////////
+
+const IR::Node* ReplaceHeaders::preorder(IR::P4Program* program) {
+    if (replacementMap->empty()) {
+        // nothing to do
+        prune();
+    }
+    return program;
+}
+
+const IR::Node* ReplaceHeaders::postorder(IR::Type_Header* type) {
+    auto canon = replacementMap->typeMap->getTypeType(getOriginal(), true);
+    auto repl = replacementMap->getReplacement(canon);
+    if (repl != nullptr)
+        return repl->replacementType;
+    return type;
+}
+
+const IR::Node* ReplaceHeaders::postorder(IR::Member* expression) {
+    // Find out if this applies to one of the parameters that are being replaced.
+    if (getParent<IR::Member>() != nullptr)
+        // We only want to process the outermost Member
+        return expression;
+
+    // this is a reference to whole header str.hdr
+    if (expression->expr->is<IR::PathExpression>())
+        return expression;
+
+    const IR::Expression* e = expression;
+    cstring prefix = "";
+    // param structs have been flattend, but we can have param.header.struct.field
+    while (auto mem = e->to<IR::Member>()) {
+        if (mem->expr->is<IR::PathExpression>()) break;
+        e = mem->expr;
+        prefix = cstring(".") + mem->member + prefix;
+    }
+    auto repl = ::get(toReplace, e->toString());
+    if (repl == nullptr)
+        return expression;
+    auto newFieldName = ::get(repl->fieldNameRemap, prefix);
+    const IR::Expression* result;
+    if (newFieldName.isNullOrEmpty()) {
+        BUG("cannot find replacement for %s", prefix);
+        /*
+        std::cout << "expression = " << expression << "\n";
+        std::cout << "e = " << e << "\n";
+        std::cout << "e->toString = " << e->toString() << "\n";
+        std::cout << "repl = " << repl << "\n";
+        std::cout << "prefix = " << prefix << "\n";
+        std::cout << "newFieldName = " << newFieldName << "\n";
+        std::cout << "cannot find replacement for " << prefix << "\n";
+        return expression;
+        */
+        //  result = repl->explode(e, prefix);
+    } else {
+        result = new IR::Member(e, newFieldName);
+    }
+    LOG3("Replacing " << expression << " with " << result);
+    return result;
+}
+
+const IR::Node* ReplaceHeaders::preorder(IR::P4Parser* parser) {
+    for (auto p : parser->getApplyParameters()->parameters) {
```

Comment:
FindTypesToReplace should have already added this type to the replacement map.

Reply:
If not, you should modify FindTypesToReplace to do that.

Reply:
It's needed. FindTypesToReplace does find headers with structs and them and replacement type with flattened struct is generated.

But this part deals with field references when the header type is used inside of a struct that is used as argument to control. Consider 
```
struct alt_t {
  bit<1> valid;
  bit<7> port;
};

struct row_t {
  alt_t alt0;
  alt_t alt1;
};

header bitvec_hdr {
  row_t row;
}
struct parsed_packet_t {
    bitvec_hdr bvh0;
    bitvec_hdr bvh1;   
};
control ingress(inout parsed_packet_t h ) {
        h.bvh0.row.alt0.valid = 0;
}
```
This code find that within control block name h.bvh0 & h.bvh1 are used to refer to header of type bitvec_hdr, which is header with struct inside, so a flattened replacement for that header exists. Therefore, 'h.bvh0' & 'h.bvh1' are added to the toReplace map.

So when a reference such as h.bvh0.row.alt0.valid is found, we split it into two parts: "h.bvh0" and ".row.alt0.valid". First part is used as the key into toReplace map, and get the replacement type. 2nd part is used as the key into fieldmap for the returned flattened header type to get the name of the flattened field.

Reply:
Note that the 'postorder(IR::Member* expression)' methods are different for ReplaceHeaders vs. ReplaceStructs.

Reply:
There are two things that could be done: 
- use the same method but handle two cases
- have two passes as there are now, but mark all the structs in the same FindTypesToReplace pass. You can keep them in two different maps if you want.

It may be actually simpler to write a new complete pass instead of inserting your code in FlattenInterfaceStructs. The reason is that you probably want both these transformations applied in sequence, and not at the same time (the typeMap will have to be recomputed between the two; both of them are changing the types of parameters in different ways). 

So it probably would be simpler to create a pass FlattenInterfaceHeaders with a similar structure and move your code there; run it after FlattenInterfaceStructs. But it should also not scan the parameters, instead it should recursively traverse the package argument types and find fields that are headers that need to be flattened.

Reply:
@mbudiu-vmw : We developed a new pass called FlattenHeaders and unit tested it to work.  Since our testing is not complete, we didn't check in the code to this PR.  One can see the code at this public URL:

https://gitlab.com/hemant_mnkcg/p4c-1

See p4c/midend/flattenHeaders.h[.cpp]

  We are calling this pass before the FlattenInterfaceStructs pass because the FlattenInterfaceStructs pass doesn't expect structs within headers.  Headers have to be flattened first.  

For more testing we created the data structures below.

`// header with only bits.
header bonly {
  bit<8> x;
  bit<24> y;
}
// nested structs are defined in p4c/testdata/p4_16_samples/issue383-bmv2.p4
struct local_metadata_t {
  row_t row0; // nested struct 
  row_t row1; // nested struct
  col_t col; // nested struct
  bonly bo;
};
`

The local_metadata_t as defined above causes the FlattenInterfaceStructs pass to fail.  The code assumes that any nested reference param.struct.field.foo.bar.moo is a nested struct and calculates flattened prefix as struct_field_foo_bar_moo.  But struct could have header or header_union in it.  In that case prefix is wrong and replacement would fail.  

The other issue is that if local_metadata_t is used with resubmit and clone, we read Andy's note at this URL:

https://github.com/jafingerhut/p4-guide/blob/master/v1model-special-ops/v1model-special-ops.p4

See lines 218-235.  Metadata should not contain a struct if the metadata is being used by resubmit/clone.  Should p4c fail a metadata used by resubmit/clone if the metadata includes a struct?

Reply:
If you found a correct program that does not compile correctly then you should file an issue for it.
We have multiple compilers, with different constraints. The bmv2 compiler should reject programs the architecture cannot handle.

Reply:
Filed an Issue:

https://github.com/p4lang/p4c/issues/1642

Reply:
We moved flattening of nested struct(s) inside a P4 header to a new midend pass, viz., FlattenHeaders.  Please see the new commit few moments back.  issue383-bmv2.p4 is the test program.

Reply:
Are you implying that the PR is ready for review?

Reply:
Let me know when everything is ready. There is no point in reviewing this several times; we should do it when everything you planned to do in this PR is ready. The tests are especially important, since by reading the generated code you can understand much easier what the passes do.

Reply:
@mbudiu-vmw :

1. I added header union and header stack tests to the PR today.  Header union tested isValid() as well.
2. In the past we already added a table key using nested struct/header and also deeply nested header in struct such as struct->struct>header->struct->struct.  The deeply nested struct tests one pass removing nested header and the second pass removing nested struct.  
3. We also tested a header with bits-only in struct which found issue 1642.  
4. We have also added copying of input header to local variable in the past. 
5. We have punted SerNum and Bool in struct for a subsequent PR where we plan to add Cast support.

Any other tests?

Reply:
One thing that would be very useful is to have a test that uses a header or a header field as a table key, accompanied by an STF file to certify that the code generated for the control-plane API works correctly.

Reply:
Will develop this test, thanks.  I haven't used a STF file yet for any test, but let me look into it since some tests use a STF file in p4c/testdata/p4_16_samples.

Reply:
A new test defined a P4 header with a nested struct inside.  This P4 header is used for a key to a table.  The STF test was developed and passed testing.  New test's files are in
p4c/testdata/p4_16_samples/bvec-hdr-bmv2.p4[.stf]

I also modified issue383-bmv2.p4 to add an Annotation to a data structure and testing passed.  Code is checked in.

There is another test I am looking at.  I will describe this test later in my day.

Reply:
This is the test that will fail and needs investigation. 

Simply, define the following struct/header:
struct str {
  bit<8> a;
  bit<8> b;
};

header parsed_packet_t {
  str s;
  bit<16> c;
}; 

Then use the header with clone and emit as follows:

clone3(CloneType.I2E, 0, h.s);
 b.emit(h.s);
 
Flattening of references will flatten h.s as a list.  But if the same flattening is applied to emit, it is wrong.  I don't believe emit can accept a list?  This issue has impact on using explode() or not in the flattening passes.   Right now, the P4 program does not even compile and fails compiling 'b.emit(h.s)' with this error:

./testdata/p4_16_samples/foo.p4(81): error: b.emit: argument must be a header, stack or union, or a struct or tuple of such types
        b.emit(h.s);
        ^^^^^^^^^^^
Clearly, checkEmitTypes() in typeChecker.cpp needs work because 'h.s' is a struct.  The p4c used is from this PR branch.